gemma/gemini_api_server.cc

```diff
@@ -0,0 +1,510 @@
+      {"outputTokenLimit", 8192},
+      {"supportedGenerationMethods", json::array({"generateContent", "streamGenerateContent"})},
+      {"temperature", 1.0},
+      {"topK", 1}
+    }}}
+  };
+  
+  res.set_content(response.dump(), "application/json");
+}
+
+// void HandleShutdown(int signal) {
+//   std::cerr << "\nShutting down server..." << std::endl;
+//   server_running = false;
+// }
+
+void RunServer(const LoaderArgs& loader, const ThreadingArgs& threading, 
+               const InferenceArgs& inference, int port) {
+  std::cerr << "Loading model..." << std::endl;
+  
+  // Initialize model
+  ThreadingContext ctx(threading);
+  MatMulEnv env(ctx);
+  
+  ServerState state;
+  state.gemma = std::make_unique<Gemma>(loader, inference, ctx);
+  state.env = &env;
+  state.ctx = &ctx;
+  
+  httplib::Server server;
+  
+  // Set up routes
+  server.Get("/", [](const httplib::Request&, httplib::Response& res) {
+    res.set_content("Gemini API Server (gemma.cpp) - Use POST /v1beta/models/gemma-2b:generateContent", "text/plain");
+  });
+  
+  // Gemini API endpoints
+  server.Get("/v1beta/models", [&state](const httplib::Request& req, httplib::Response& res) {
+    HandleListModels(state, req, res);
+  });
+  
+  server.Post("/v1beta/models/gemma-2b:generateContent", [&state](const httplib::Request& req, httplib::Response& res) {
+    HandleGenerateContentNonStreaming(state, req, res);
+  });
+  
+  server.Post("/v1beta/models/gemma-2b:streamGenerateContent", [&state](const httplib::Request& req, httplib::Response& res) {
+    HandleGenerateContentStreaming(state, req, res);
+  });
+  
+  // Periodic cleanup of old sessions
+  std::thread cleanup_thread([&state]() {
+    while (server_running) {
+      std::this_thread::sleep_for(std::chrono::minutes(5));
+      state.CleanupOldSessions();
+    }
+  });
+  
+  std::cerr << "Starting Gemini API server on port " << port << std::endl;
+  std::cerr << "Model loaded successfully" << std::endl;
+  std::cerr << "Endpoints:" << std::endl;
+  std::cerr << "  POST /v1beta/models/gemma-2b:generateContent" << std::endl;
+  std::cerr << "  POST /v1beta/models/gemma-2b:streamGenerateContent (SSE)" << std::endl;
+  std::cerr << "  GET  /v1beta/models" << std::endl;
+  
+  if (!server.listen("0.0.0.0", port)) {
+    std::cerr << "Failed to start server on port " << port << std::endl;
+  }
+  
+  cleanup_thread.join();
+}
+
+}  // namespace gcpp
+
+int main(int argc, char** argv) {
+  gcpp::InternalInit();
+  
+  gcpp::LoaderArgs loader(argc, argv);
+  gcpp::ThreadingArgs threading(argc, argv);
+  gcpp::InferenceArgs inference(argc, argv);
+  
+  if (gcpp::HasHelp(argc, argv)) {
+    std::cerr << "\n\nGemini API server for gemma.cpp\n";
+    std::cerr << "Usage: " << argv[0] << " --weights <path> --tokenizer <path> [options]\n";
+    std::cerr << "\nOptions:\n";
+    std::cerr << "  --port PORT      Server port (default: 8080)\n";
+    std::cerr << "\n";
+    return 0;
+  }
+  
+  // Parse port
```

Comment:
If you want, you can also add a port member to InferenceArgs, with a comment that it is only used in the api_server.cc. That way, your new argument will benefit from any updates to arg parsing/handling.

Reply:
Added port and model to InferenceArgs.